src/common/helpers/validation-warnings.js

```diff
@@ -194,6 +196,57 @@ export const generateHazardousConsignmentWarnings = (payload) => {
   return warnings
 }
 
+/**
+ * Generate warnings for source of component conditions
+ * If source of components is provided then hazardous components must be provided, otherwise add a warning
+ *
+ * @param {Object} payload
+ * @returns {Array}
+ */
+export const generateSourceOfComponentsWarnings = (payload) => {
+  const warnings = []
+
+  if (!payload?.waste?.hazardous) {
+    return warnings
+  }
+
+  const sourceOfComponents = payload.waste.hazardous.sourceOfComponents
+  const hazardousComponents = payload.waste.hazardous.components
+
+  // If source of components has been provided then all hazardous components should have a name and concentration
+  if (
+    Object.keys(sourceOfComponentsProvided).includes(sourceOfComponents) &&
+    !haveAllHazardousComponentsGotNameAndConcentration(hazardousComponents)
```

Comment:
haveAllHazardousComponentsGotNameAndConcentration treats any falsy concentration as missing. A concentration of 0 is allowed by the schema but will be flagged incorrectly, so every 0% component will now raise a warning. -> I think js treats 0 as falsy